storage/local/mapper.go

```diff
@@ -0,0 +1,179 @@
+package local
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/golang/glog"
+
+	clientmodel "github.com/prometheus/client_golang/model"
+)
+
+const maxMappedFP = 1 << 30 // About 1B fingerprints reserved for mapping.
+
+var separatorString = string([]byte{clientmodel.SeparatorByte})
+
+// fpMappings maps original fingerprints to a map of string representations of
+// metrics to the truly unique fingerprint.
+type fpMappings map[clientmodel.Fingerprint]map[string]clientmodel.Fingerprint
+
+// fpMapper is used to map fingerprints in order to work around fingerprint
+// collisions.
+type fpMapper struct {
+	mtx      sync.RWMutex // Protects collisions.
+	mappings fpMappings
+
+	fpToSeries      *seriesMap
+	p               *persistence
+	highestMappedFP clientmodel.Fingerprint
+}
+
+// newFPMapper loads the collision map from the persistence and
+// returns an fpCollisionResolver ready to use.
```

Comment:
I named everything something-collision-something at first, but that proved to be confusing, so I reverted to mapping/mapper (and mention in the doc string that the mapping help to avoid collisions).